remote_test.go

```diff
@@ -1555,3 +1559,140 @@ func (s *RemoteSuite) TestFetchAfterShallowClone(c *C) {
 		plumbing.NewSymbolicReference("HEAD", "refs/heads/master"),
 	})
 }
+
+func TestFetchFastForwardForCustomRef(t *testing.T) {
+	customRef := "refs/custom/branch"
+	// 1. Set up a remote with a URL
+	remoteURL := t.TempDir()
+	remoteRepo, err := PlainInit(remoteURL, true)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// 2. Add a commit with an empty tree to master and custom ref, also set HEAD
+	emptyTreeID := writeEmptyTree(t, remoteRepo)
+	writeCommitToRef(t, remoteRepo, "refs/heads/master", emptyTreeID)
+	writeCommitToRef(t, remoteRepo, customRef, emptyTreeID)
+	if err := remoteRepo.Storer.SetReference(plumbing.NewSymbolicReference(plumbing.HEAD, "refs/heads/master")); err != nil {
+		t.Fatal(err)
+	}
+
+	// 3. Clone repo, then fetch the custom ref
+	// Note that using custom ref in ReferenceName has an IsBranch issue
+	localRepo, err := Clone(memory.NewStorage(), memfs.New(), &CloneOptions{
+		URL: remoteURL,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := localRepo.Fetch(&FetchOptions{
+		RefSpecs: []config.RefSpec{
+			config.RefSpec(fmt.Sprintf("%s:%s", customRef, customRef)),
+		},
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	// 4. Make divergent changes
+	remoteCommitID := writeCommitToRef(t, remoteRepo, customRef, emptyTreeID)
+	// Sleep so that consecutive calls to writeCommitToRef have different
+	// time.Now() values
+	time.Sleep(time.Second)
```

Comment:
Instead of sleeping, we could send time as a parameter for `writeCommitToRef`.

```suggestion
```

Reply:
Done!